### PR TITLE
Add validation on Application/Runtime name in graphql mutations

### PR DESCRIPTION
**Pull request by dbadura**
**Description**

Changes proposed in this pull request:

- Add validation on application and runtime name when creating or updating. Name has to be in accordance with [Kubernetes Naming Convention](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names)
- Add validation that name of application or runtime is unique.

**Related issue(s)**

See #135

### CODE REVIEW

File: components/director/internal/domain/application/service.go [hunk 1]
```diff
@@ -267,3 +285,16 @@ func (s *service) deleteRelatedResources(applicationID string) error {
 
 	return nil
 }
+
+func validateApplicationName(name string) (bool, error) {
+	match, err := regexp.MatchString("^[a-z\\.-]+$", name)
```
Comment (aszecowka) [inline components/director/internal/domain/application/service.go]: your regexp does not match what is described here: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
for example max length of name. 
Please check if you can use validation from k8s itself: https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/apps/validation/validation.go#L379

File: components/director/internal/domain/application/service.go [hunk 2]
```diff
@@ -125,11 +135,19 @@ func (s *service) Create(ctx context.Context, in model.ApplicationInput) (string
 }
 
 func (s *service) Update(ctx context.Context, id string, in model.ApplicationInput) error {
+	valid, err := validateApplicationName(in.Name)
```
Comment (aszecowka) [inline components/director/internal/domain/application/service.go]: maybe we can add Validate method on ApplicationInput?

File: components/director/internal/domain/application/service_test.go [hunk 3]
```diff
@@ -140,7 +198,11 @@ func TestService_Create(t *testing.T) {
 
 			// then
 			assert.IsType(t, "string", result)
-			assert.Equal(t, testCase.ExpectedErr, err)
+			if err == nil {
```
Comment (aszecowka) [inline components/director/internal/domain/application/service_test.go]: try to simplify it with assert.Equal

File: components/director/internal/model/runtime_test.go [hunk 4]
```diff
@@ -200,3 +200,49 @@ func TestRuntimeInput_ToRuntime(t *testing.T) {
...
+	}
+
+	for i, testCase := range testCases {
+		t.Run(fmt.Sprintf("%d: %s", i, testCase.Name), func(t *testing.T) {
+			//WHEN
+			errorMsg := testCase.Input.ValidateInput()
+
+			//THEN
+			if testCase.ExpectedErrMsg != nil {
+				require.Equal(t, 1, len(errorMsg))
```
Comment (aszecowka) [inline components/director/internal/model/runtime_test.go]: require.Len

File: components/director/internal/model/runtime_test.go [hunk 5]
```diff
@@ -200,3 +200,49 @@ func TestRuntimeInput_ToRuntime(t *testing.T) {
...
 	}
 }
+
+func TestRuntimeInput_ValidateInput(t *testing.T) {
+	//GIVEN
+	validationErrorMsg := []string{"a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character"}
+	testCases := []struct {
+		Name           string
+		Input          model.RuntimeInput
+		ExpectedErrMsg []string
```
Comment (aszecowka) [inline components/director/internal/model/runtime_test.go]: try to change it to expectedErr of type error

File: components/director/internal/model/runtime.go [hunk 6]
```diff
@@ -94,6 +96,10 @@ func (i *RuntimeInput) ToRuntime(id string, tenant string) *Runtime {
 	}
 }
 
+func (i *RuntimeInput) ValidateInput() []string {
```
[the remark this answers is not in the document]
Comment (dbadura) [reply components/director/internal/model/runtime.go]: I decided to return error with all error messages as text inside.

File: components/director/internal/domain/application/service_test.go (hunk 3 above)
Comment (dbadura) [reply components/director/internal/domain/application/service_test.go]: The basic problem with k8s validation err messages Is that they contain very long message with regex inside.

File: components/director/internal/domain/runtime/repository.go [hunk 7]
```diff
@@ -60,11 +65,21 @@ func (r *inMemoryRepository) Update(item *model.Runtime) error {
...
 
-	r.store[item.ID] = item
+	if rtm.Name != item.Name {
+		found := r.findRuntimeNameWithinTenant(item.Tenant, item.Name)
+		if found {
+			return errors.New("Runtime name is not unique within tenant")
+		}
+	}
+
+	rtm.Name = item.Name
```
Comment (aszecowka) [inline components/director/internal/domain/runtime/repository.go]: why not `r.store[item.ID]=item` ?

File: components/director/internal/domain/runtime/service_test.go [hunk 8]
```diff
@@ -58,6 +58,31 @@ func TestService_Create(t *testing.T) {
 			Input:       modelInput,
 			ExpectedErr: nil,
 		},
+		{
+			Name: "Returns error when name is empty",
```
Comment (aszecowka) [inline components/director/internal/domain/runtime/service_test.go]: you test the same thing twice: in model package and here. Here you should have a test that checks if an input is validated - one scenario will be enough.

File: components/director/internal/domain/runtime/repository.go [hunk 9]
```diff
@@ -60,8 +65,16 @@ func (r *inMemoryRepository) Update(item *model.Runtime) error {
...
 	}
 
-	if r.store[item.ID] == nil {
-		return errors.New("application not found")
+	oldRuntine := r.store[item.ID]
+	if oldRuntine == nil {
+		return errors.New("Runtime not found")
+	}
+
+	if oldRuntine.Name != item.Name {
```
[the remark this answers is not in the document]
Comment (dbadura) [reply components/director/internal/domain/runtime/repository.go]: We don't have repository tests.

File: tests/end-to-end/director/application_api_test.go [hunk 10]
```diff
@@ -392,12 +393,100 @@ func TestUpdateApplication(t *testing.T) {
...
 
+	//THEN
+	require.NoError(t, err)
 	assertApplication(t, in, updatedApp)
 }
 
+func TestCreateUpdateApplicationWithDuplicatedNamesWithinTenant(t *testing.T) {
+	// GIVEN
+	ctx := context.Background()
+	in := generateSampleApplicationInput("app")
```
Comment (aszecowka) [inline tests/end-to-end/director/application_api_test.go]: generateSampleApplicationInput already set application name

File: tests/end-to-end/director/application_api_test.go [hunk 11]
```diff
@@ -392,12 +393,100 @@ func TestUpdateApplication(t *testing.T) {
...
+	err = tc.RunQuery(ctx, createReq, &firstApp)
+
+	//THEN
+	require.NoError(t, err)
+	require.NotEmpty(t, firstApp.ID)
+	defer deleteApplication(t, firstApp.ID)
+
+	//Create second application with first application name
+	//GIVEN
+	in = generateSampleApplicationInput("app")
```
Comment (aszecowka) [inline tests/end-to-end/director/application_api_test.go]: in was previously created, not need to do it one more time

File: tests/end-to-end/director/application_api_test.go [hunk 12]
```diff
@@ -392,12 +393,100 @@ func TestUpdateApplication(t *testing.T) {
...
+	err = tc.RunQuery(ctx, createReq, nil)
+
+	//THEN
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "Application name is not unique within tenant")
+
+	//Create second application with unique name
+	//GIVEN
+	in = generateSampleApplicationInput("app")
+	in.Name = "application-2"
```
Comment (aszecowka) [inline tests/end-to-end/director/application_api_test.go]: see above comments

File: tests/end-to-end/director/application_api_test.go [hunk 13]
```diff
@@ -392,12 +393,100 @@ func TestUpdateApplication(t *testing.T) {
...
+	err = tc.RunQuery(ctx, createReq, &secondApp)
+
+	//THEN
+	require.NoError(t, err)
+	require.NotEmpty(t, secondApp.ID)
+	defer deleteApplication(t, secondApp.ID)
+
+	// Try to update second app with name from first app
+	//GIVEN
+	in = generateSampleApplicationInput("app")
```
Comment (aszecowka) [inline tests/end-to-end/director/application_api_test.go]: in was already created, not need to do it one more time

File: tests/end-to-end/director/runtime_api_test.go [hunk 14]
```diff
@@ -113,6 +175,111 @@ func TestRuntimeCreateUpdateAndDelete(t *testing.T) {
 	require.NoError(t, err)
 }
 
+func TestRuntimeCreateUpdateDuplicatedNames(t *testing.T) {
```
Comment (aszecowka) [inline tests/end-to-end/director/runtime_api_test.go]: TestRuntimeNameValidation maybe will be better name?

**Pull request merged**